core/src/main/java/org/jclouds/reflect/Reflection2.java

```diff
@@ -153,15 +153,23 @@
          .newBuilder().build(new CacheLoader<TypeToken<?>, Set<Invokable<?, ?>>>() {
             public Set<Invokable<?, ?>> load(TypeToken<?> key) {
                ImmutableSet.Builder<Invokable<?, ?>> builder = ImmutableSet.<Invokable<?, ?>> builder();
-               for (Constructor<?> ctor : key.getRawType().getDeclaredConstructors()) {
-                  ctor.setAccessible(true);
+               Class<?> raw = key.getRawType();
+               for (Constructor<?> ctor : raw.getDeclaredConstructors()) {
+                  if (!coreJavaClass(raw)) {
+                     // In JDK 11 up to 14, the only uses for `java.beans.ConstructorProperties` annotation
+                     // are in the `java.awt`, `java.beans` and `javax.swing` packages.
+                     // Since these constructors are public, there is no need to call `setAccessible()`
```

Comment:
Instead of doing this, could we just do `if (!ctor.isAccessible()) ctor.setAccessible(true)`? Probably this is better for the general case and makes the code safer (and this comment irrelevant)?
We can keep the core java class checks if needed, but it looks a bit safer to just check if it is accessible, then try setting it. And if not accessible and a core Java class, we can probably log a warning message as something is probably gonna fail later as we can't set the proper access level tot he method?

Reply:
Given the javadoc of  [`AccessibleObject.isAccessible()`](https://docs.oracle.com/javase/9/docs/api/java/lang/reflect/AccessibleObject.html#isAccessible--), it looks almost certain that `ctor.isAccessible()` will return false for that package private constructor of String, and so `ctor.setAccessible(true)` will be called anyway and we will get the illegal access warning again. Our code is not calling  `ctor.setAccessible(true)` at all.

Even when using [`AccessibleObject.canAccess()`](https://docs.oracle.com/javase/9/docs/api/java/lang/reflect/AccessibleObject.html#canAccess-java.lang.Object-), it will return false in our case and code will end up calling `ctor.setAccessible(true)`, thus we will still get the illegal access warning.

Am I missing or misunderstanding something?

Reply:
I read again the code and your comment, and I think you are saying I could drop the last sentence of the comment and then do the following?
```java
                  if (!ctor.canAccess(null) && !coreJavaClass(raw)) {
                     // In JDK 11 up to 14, the only uses for `java.beans.ConstructorProperties` annotation
                     // are in the `java.awt`, `java.beans` and `javax.swing` packages.
                     ctor.setAccessible(true);
                  }
```
We need to test it, but that should work indeed. There would be less spurious warnings for regular cases.

The check on `coreJavaClass()` is still required in our case though.

Reply:
> Instead of doing this, could we just do if (!ctor.isAccessible()) ctor.setAccessible(true)?

@nacx That would indeed be cleaner, but I suspect (as I think @JnRouvignac is describing) this will not fix the warnings:

* jclouds explicitly tries to make certain non-accessible members accessible via reflection, and needs this to work
* **some** of those non-accessible members can be made accessible, **some** trigger warnings
* the latter set contains core classes that jclouds actually does **not** need to make accessible, so it should be safe to skip them

I no longer recall why we didn't add the "skip core classes" check to `constructorsForTypeToken` when we added it to `methodsForTypeToken` in db4e4af (I suspect it wasn't required to fix whatever issue we were seeing then?); it seems consistent enough to me to add it now, assuming all tests pass.

I agree that your suggestion - only try to make members accessible that aren't *already* accessible - would be a logic improvement, as we would only try to modify members where needed. I wonder, though, if it may actually be less performant, given that `isAccessible` would have to be invoked many times?

I think it makes sense to consider that potential improvement - here and also in `methodsForTypeToken` - as a separate PR if desired, though.

Happy to discuss further if useful!

Reply:
Note 1: `isAccessible()` is not useful. I think you meant `canAccess()`.
Note 2: if if understand correctly, this code does not look performance sensitive: it is only called once for each `TypeToken` (i.e. for each class?) in the lifetime of an application, since I see no expiration for this cache.

> I think it makes sense to consider that potential improvement - here and also in methodsForTypeToken - as a separate PR if desired, though.

I would also prefer to have the original PR in as it fixes the problem, then I am also happy to follow-up with another to add the `canAccess()` checks everywhere.
Let me know how you want to proceed.

Reply:
Thanks for the replies, and yes, that makes sense. Let's keep the original proposed change as it addresses the issue and it aligns with the existing approach in the other method Andrew mentioned.